R/bs-theme-preset-brand.R

```diff
@@ -0,0 +1,830 @@
+      )
+    }
+
+    font_data_uri <- if (grepl("^https?://", file$path)) {
+      font_path <- file$path
+    } else {
+      font_path <- file.path(brand_root, file$path)
+      base64enc::dataURI(
+        file = font_path,
+        mime = mime::guess_type(font_path)
+      )
+    }
+    font_type <- switch(
+      tools::file_ext(tolower(font_path)),
+      # otc = "collection",
+      # ttc = "collection",
+      # eot = "embedded-opentype",
+      otf = "opentype",
+      ttf = "truetype",
+      # svg = "svg",
+      # svgz = "svg",
+      woff = "woff",
+      woff2 = "woff2",
+      abort(
+        c(
+          sprintf("Invalid font type: %s", font_path),
+          "i" = "Font must be `.ttf`, `.otf`, `.woff` or `.woff2`."
+        )
+      )
+    )
+
+    sass::font_face(
+      family = family,
+      src = sprintf("url(%s) format(%s)", font_data_uri, font_type),
+      weight = brand_remap_font_weight(file$weight),
+      style = file$style,
+      display = "auto"
+    )
+  })
+
+  sass::font_collection(!!!font_collection_files)
+}
+
+brand_remap_font_weight <- function(x) {
+  if (is.null(x)) return()
+
+  for (i in seq_along(x)) {
+    if (x[[i]] %in% names(brand_font_weight_map)) {
+      x[[i]] <- brand_font_weight_map[x[[i]]]
+    }
+  }
+
+  x
+}
+
+brand_font_weight_map <- c(
+  "thin" = 100,
+  "extra-light" = 200,
+  "ultra-light" = 200,
+  "light" = 300,
+  "normal" = 400,
+  "regular" = 400,
+  "medium" = 500,
+  "semi-bold" = 600,
+  "demi-bold" = 600,
+  "bold" = 700,
+  "extra-bold" = 800,
+  "ultra-bold" = 800,
+  "black" = 900
+)
+
+#' Convert a font size to rem
+#'
+#' Some frameworks, like Bootstrap expect base font size to be in `rem`. This
+#' function converts `em`, `%`, `px`, `pt` to `rem`:
+#'
+#' 1. `em` is directly replace with `rem`.
+#' 2. `1%` is `0.01rem`, e.g. `90%` becomes `0.9rem`.
+#' 3. `16px` is `1rem`, e.g. `18px` becomes `1.125rem`.
+#' 4. `12pt` is `1rem`.
+#' 5. `0.1666in` is `1rem`.
+#' 6. `4.234cm` is `1rem`.
+#' 7. `42.3mm` is `1rem`.
+#'
+#' @noRd
+maybe_convert_font_size_to_rem <- function(x) {
+  x_og <- as.character(x)
+  split_result <- split_css_value_and_unit(x)
+  value <- split_result$value
+  unit <- split_result$unit
+
+  if (unit %in% c("rem", "em")) {
+    return(paste0(value, "rem"))
+  }
+
+  scale <- list(
+    "%" = 100,
+    "px" = 16,
+    "pt" = 12,
+    "in" = 16 / 96, # 96 px/inch
+    "cm" = 16 / 96 * 2.54, # inch -> cm
+    "mm" = 16 / 96 * 25.4 # cm -> mm
+  )
+
+  if (unit %in% names(scale)) {
+    return(paste0(as.numeric(value) / scale[[unit]], "rem"))
+  }
+
+  if (unit == "") {
+    unit <- "unknown"
+  }
+
+  stop(
+    paste0(
+      "Could not convert font size '",
+      x_og,
+      "' from ",
+      unit,
+      " units to a relative unit."
+    )
+  )
+}
+
+split_css_value_and_unit <- function(x) {
+  pattern <- "^(-?[0-9]*\\.?[0-9]+)\\s*([a-z%]*)$"
+  match <- regexec(pattern, x)
+  result <- regmatches(x, match)[[1]]
```

Comment:
Does this need to be concerned with leading/trailing whitespace?

Reply:
Also might be worth handling `calc()` more gracefully

Reply:
I agree that gracefully handling `calc()` is a good idea in general, but in the specific use case and context (root font size setting for brand.yml) we don't want to encourage or support its use.